mongodbatlas/fw_resource_mongodbatlas_database_user.go

```diff
@@ -0,0 +1,529 @@
+	DatabaseName   types.String `tfsdk:"database_name"`
+}
+
+type tfLabelModel struct {
+	Key   types.String `tfsdk:"key"`
+	Value types.String `tfsdk:"value"`
+}
+
+type tfScopeModel struct {
+	Name types.String `tfsdk:"name"`
+	Type types.String `tfsdk:"type"`
+}
+
+type DatabaseUserRS struct {
+	client *MongoDBClient
+}
+
+var RoleObjectType = types.ObjectType{AttrTypes: map[string]attr.Type{
+	"role_name":       types.StringType,
+	"collection_name": types.StringType,
+	"database_name":   types.StringType,
+}}
+
+var LabelObjectType = types.ObjectType{AttrTypes: map[string]attr.Type{
+	"key":   types.StringType,
+	"value": types.StringType,
+}}
+
+var ScopeObjectType = types.ObjectType{AttrTypes: map[string]attr.Type{
+	"name": types.StringType,
+	"type": types.StringType,
+}}
+
+func (r *DatabaseUserRS) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"project_id": schema.StringAttribute{
+				Required: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"auth_database_name": schema.StringAttribute{
+				Required: true,
+			},
+			"username": schema.StringAttribute{
+				Required: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"password": schema.StringAttribute{
+				Optional:  true,
+				Sensitive: true,
+				Validators: []validator.String{
+					stringvalidator.ConflictsWith(path.Expressions{
+						path.MatchRelative().AtParent().AtName("x509_type"),
+						path.MatchRelative().AtParent().AtName("ldap_auth_type"),
+						path.MatchRelative().AtParent().AtName("aws_iam_type"),
+					}...),
+				},
+			},
+			"x509_type": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				Default:  stringdefault.StaticString("NONE"),
+				Validators: []validator.String{
+					stringvalidator.OneOf("NONE", "MANAGED", "CUSTOMER"),
+				},
+			},
+			"oidc_auth_type": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				Default:  stringdefault.StaticString("NONE"),
+				Validators: []validator.String{
+					stringvalidator.OneOf("NONE", "IDP_GROUP"),
+				},
+			},
+			"ldap_auth_type": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				Default:  stringdefault.StaticString("NONE"),
+				Validators: []validator.String{
+					stringvalidator.OneOf("NONE", "USER", "GROUP"),
+				},
+			},
+			"aws_iam_type": schema.StringAttribute{
+				Optional: true,
+				Computed: true,
+				Default:  stringdefault.StaticString("NONE"),
+				Validators: []validator.String{
+					stringvalidator.OneOf("NONE", "USER", "ROLE"),
+				},
+			},
+		},
+		Blocks: map[string]schema.Block{
+			"roles": schema.SetNestedBlock{
```

Comment:
`roles` is a required attribute today, how can we ensure that is maintained? Blocks are optional by default

Also by looking at the docs for this resource here https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/database_user#roles 
looks like we expect role_name and database_name to be required as well. However, I see that they are optional even today. 
Can we confirm these attribute requirements?

Reply:
> roles is a required attribute today, how can we ensure that is maintained? Blocks are optional by default

Thanks. I missed it. This was easier than expected, you just need to add a validator

```
			"roles": schema.SetNestedBlock{
				Validators: []validator.Set{
					setvalidator.IsRequired(),
				},
```

I tested locally and this is the error that you get if you do not provide `roles`
```
╷
│ Error: Invalid Block
│
│   with mongodbatlas_database_user.test2,
│   on main.tf line 615, in resource "mongodbatlas_database_user" "test2":
│  615: resource "mongodbatlas_database_user" "test2" {
│
│ Block roles must have a configuration value as the provider has marked it as required
```

I will update the wiki what I learned if this is missing. Thanks

> Also by looking at the docs for this resource here https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/database_user#roles
looks like we expect role_name and database_name to be required as well. However, I see that they are optional even today. Can we confirm these attribute requirements?

good point. Yes, I checked the open API doc and they are indeed required. I updated the schema.

Reply:
Thank you!